File: GO/resolve_optimized.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"gonum.org/v1/gonum/mat"
)

func main() {
	fmt.Println("[INFO] Chargement des matrices depuis matrices.json...")

	f, err := os.Open("matrices.json")
	if err != nil {
		fmt.Println("Erreur lors de l'ouverture du fichier :", err)
		return
	}
	defer f.Close()

	var data map[string][][]float64
	if err := json.NewDecoder(f).Decode(&data); err != nil {
		fmt.Println("Erreur lors du décodage JSON :", err)
		return
	}

	a := data["A"]
	b := data["B"]

	rows, cols := len(a), len(b[0])
	n := len(a[0]) // nombre de colonnes de A = nombre de lignes de B

	flatA := make([]float64, 0, rows*n)
	for _, row := range a {
		flatA = append(flatA, row...)
	}

	flatB := make([]float64, 0, n*cols)
	for _, row := range b {
		flatB = append(flatB, row...)
	}

	matA := mat.NewDense(rows, n, flatA)
	matB := mat.NewDense(n, cols, flatB)

	fmt.Println("[INFO] Multiplication avec gonum...")
	start := time.Now()
	var result mat.Dense
	result.Mul(matA, matB)
	elapsed := time.Since(start)

	fmt.Printf("[RESULT] Temps de résolution (Go + gonum) : %s\n", elapsed)
}

File: C/resolve_optimized.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "cJSON.h"
#include <cblas.h>

#define SIZE 1000

double *flatten_matrix(double **matrix) {
    double *flat = malloc(SIZE * SIZE * sizeof(double));
    for (int i = 0; i < SIZE; i++)
        for (int j = 0; j < SIZE; j++)
            flat[i * SIZE + j] = matrix[i][j];
    return flat;
}

double **allocate_matrix() {
    double **matrix = malloc(SIZE * sizeof(double *));
    for (int i = 0; i < SIZE; i++) {
        matrix[i] = malloc(SIZE * sizeof(double));
    }
    return matrix;
}

void free_matrix(double **matrix) {
    for (int i = 0; i < SIZE; i++) {
        free(matrix[i]);
    }
    free(matrix);
}

double **parse_matrix(cJSON *json_matrix) {
    double **matrix = allocate_matrix();
    int i = 0;
    cJSON *row;
    cJSON_ArrayForEach(row, json_matrix) {
        int j = 0;
        cJSON *val;
        cJSON_ArrayForEach(val, row) {
            matrix[i][j++] = val->valuedouble;
        }
        i++;
    }
    return matrix;
}

int main() {
    printf("[INFO] Chargement des matrices depuis matrices.json...\n");

    FILE *f = fopen("matrices.json", "rb");
    if (!f) {
        printf("Erreur ouverture du fichier.\n");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    long len = ftell(f);
    fseek(f, 0, SEEK_SET);

    char *data = malloc(len + 1);
    fread(data, 1, len, f);
    data[len] = '\0';
    fclose(f);

    cJSON *json = cJSON_Parse(data);
    if (!json) {
        printf("Erreur JSON.\n");
        return 1;
    }

    double **A = parse_matrix(cJSON_GetObjectItem(json, "A"));
    double **B = parse_matrix(cJSON_GetObjectItem(json, "B"));
    double *flatA = flatten_matrix(A);
    double *flatB = flatten_matrix(B);
    double *result = calloc(SIZE * SIZE, sizeof(double));

    printf("[INFO] Multiplication avec OpenBLAS...\n");
    clock_t start = clock();
    cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans,
                SIZE, SIZE, SIZE, 1.0,
                flatA, SIZE,
                flatB, SIZE,
                0.0, result, SIZE);
    clock_t end = clock();

    double elapsed = (double)(end - start) / CLOCKS_PER_SEC;
    printf("[RESULT] Temps de résolution (C + OpenBLAS) : %.4f secondes\n", elapsed);

    // Clean
    free(data);
    cJSON_Delete(json);
    free_matrix(A);
    free_matrix(B);
    free(flatA);
    free(flatB);
    free(result);

    return 0;
}
